File: app/src/main/java/com/rayworks/droidweekly/SearchActivity.java
package com.rayworks.droidweekly;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.rayworks.droidweekly.model.ArticleItem;
import com.rayworks.droidweekly.model.OldItemRef;
import com.rayworks.droidweekly.repository.ArticleManager;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;

public class SearchActivity extends AppCompatActivity
        implements ArticleManager.ArticleDataListener {

    private RecyclerView recyclerView;
    private ArticleAdapter articleAdapter;
    private SearchView searchView;
    private Disposable disposable;

    public static void launch(Context context) {
        Intent intent = new Intent(context, SearchActivity.class);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_white_24dp);

        recyclerView = findViewById(R.id.result_list);
        articleAdapter = new ArticleAdapter(this, new ArrayList<>());
        articleAdapter.setViewArticleListener(
                url -> {
                    Intent intent = new Intent(Intent.ACTION_VIEW);
                    intent.setData(Uri.parse(url));
                    startActivity(intent);
                });

        recyclerView.setAdapter(articleAdapter);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.addItemDecoration(
                new DividerItemDecoration(this, layoutManager.getOrientation()));
    }

    private void setupSearchAction() {
        disposable =
                RxSearchObservable.fromView(searchView)
                        .debounce(300, TimeUnit.MILLISECONDS)
                        .filter(
                                s -> {
                                    boolean isEmpty = s.isEmpty();
                                    if (isEmpty) {
                                        resetResult();
                                    }

                                    return !isEmpty;
                                })
                        .distinctUntilChanged()
                        .switchMap(
                                new Function<String, ObservableSource<? extends String>>() {
                                    @Override
                                    public ObservableSource<? extends String> apply(String s)
                                            throws Exception {

                                        // already an async operation
                                        ArticleManager.getInstance()
                                                .search(s, getArticleListener());

                                        return Observable.just(s);
                                    }
                                })
                        .subscribe(
                                new Consumer<String>() {
                                    @Override
                                    public void accept(String s) throws Exception {
                                        System.out.println("Searching for key : " + s);
                                    }
                                });
    }

    private void resetResult() {
        runOnUiThread(
                () -> {
                    if (!SearchActivity.this.isFinishing() && recyclerView != null) {
                        articleAdapter.update(Collections.EMPTY_LIST);
                    }
                });
    }

    @NonNull
    private WeakReference<ArticleManager.ArticleDataListener> getArticleListener() {
        return new WeakReference<>(SearchActivity.this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_search, menu);
        final MenuItem item = menu.findItem(R.id.action_search);

        searchView = (SearchView) item.getActionView();
        setupSearchAction();

        // expands the SearchView automatically
        searchView.setIconified(false);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onLoadError(String err) {
        Toast.makeText(this, "Error: " + err, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onOldRefItemsLoaded(List<OldItemRef> itemRefs) {
        /** None logic * */
    }

    @Override
    public void onComplete(List<ArticleItem> items) {
        articleAdapter.update(items);
    }

    @Override
    protected void onDestroy() {
        if (disposable != null) {
            disposable.dispose();
        }

        super.onDestroy();
    }
}

File: app/src/main/java/com/rayworks/droidweekly/SearchActivity.kt
package com.rayworks.droidweekly

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.rayworks.droidweekly.databinding.ActivitySearchBinding
import com.rayworks.droidweekly.extension.scoped
import com.rayworks.droidweekly.model.ArticleItem
import com.rayworks.droidweekly.model.OldItemRef
import com.rayworks.droidweekly.repository.ArticleManager
import com.rayworks.droidweekly.repository.ArticleManager.ArticleDataListener
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import java.lang.ref.WeakReference
import java.util.*
import java.util.concurrent.TimeUnit

/***
 * The page shows the search result based on cached historical articles
 */
class SearchActivity : AppCompatActivity(), ArticleDataListener {
    private lateinit var recyclerView: RecyclerView
    private lateinit var articleAdapter: ArticleAdapter
    private lateinit var binding : ActivitySearchBinding
    private var searchView: SearchView? = null
    private var disposable: Disposable? = null


    /***
     *  A presenter handles the searching by keywords.
     */
    class SearchPresenter {
        /***
         * Search specified keywords
         */
        fun search(s: String, articleListener: WeakReference<ArticleDataListener>) {
            ArticleManager.getInstance()
                    .search(s, articleListener)
        }

        /***
         * Dispose the unused resources
         */
        fun dispose() {
            ArticleManager.getInstance().dispose()
        }
    }

    private val presenter by scoped { SearchPresenter() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySearchBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)
        val actionBar = supportActionBar
        actionBar!!.setDisplayHomeAsUpEnabled(true)
        actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_white_24dp)

        recyclerView = binding.resultList
        articleAdapter = ArticleAdapter(this, ArrayList())
        articleAdapter.setViewArticleListener { url: String? ->
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = Uri.parse(url)
            startActivity(intent)
        }

        recyclerView.apply {
            adapter = articleAdapter

            val layoutManager = LinearLayoutManager(this@SearchActivity)
            this.layoutManager = layoutManager

            addItemDecoration(
                    DividerItemDecoration(this@SearchActivity, layoutManager.orientation))
        }
    }

    private fun setupSearchAction() {
        disposable = RxSearchObservable.fromView(searchView)
                .debounce(300, TimeUnit.MILLISECONDS)
                .filter { s: String ->
                    val isEmpty = s.isEmpty()
                    if (isEmpty) {
                        resetResult()
                    }
                    !isEmpty
                }
                .distinctUntilChanged()
                .switchMap { s -> // already an async operation
                    presenter.search(s, articleListener)
                    Observable.just(s)
                }
                .subscribe { s -> println("Searching for key : $s") }
    }

    private fun resetResult() {
        runOnUiThread {
            if (!this@SearchActivity.isFinishing && recyclerView != null) {
                articleAdapter!!.update(Collections.emptyList())
            }
        }
    }

    private var articleListener: WeakReference<ArticleDataListener> = WeakReference(this@SearchActivity)

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_search, menu)
        val item = menu.findItem(R.id.action_search)
        searchView = item.actionView as SearchView
        setupSearchAction()

        // expands the SearchView automatically
        searchView!!.isIconified = false
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onLoadError(err: String) {
        Toast.makeText(this, "Error: $err", Toast.LENGTH_SHORT).show()
    }

    override fun onOldRefItemsLoaded(itemRefs: List<OldItemRef>) {
        /** None logic *  */
    }

    override fun onComplete(items: List<ArticleItem>) {
        articleAdapter!!.update(items)
    }

    override fun onDestroy() {
        if (disposable != null) {
            presenter.dispose()

            disposable!!.dispose()
        }
        super.onDestroy()
    }

    companion object {
        @JvmStatic
        fun launch(context: Context) {
            val intent = Intent(context, SearchActivity::class.java)
            context.startActivity(intent)
        }
    }
}